### DISCUSSION TITLE
Introducing ‘namespace’ for workflow definitions, task definitions, etc

### DISCUSSION THREAD

**Discussion by wonjvmware** [Netflix/conductor#2402] [Ideas]
Hi All,

I’d like to discuss the idea of introducing the concept of ‘namespace’ for workflow definitions, task definitions, etc. Currently, `name` and `version` are the components of the primary key for workflow definitions. `name` is the primary key for task definitions. 
Conductor seems to be designed to be used by one organization for its internal workflows. However, I think that it would be great if Conductor can be used by multiple organizations. Currently,  the name of a task definition is a duplicate, Conductor will return an error. If ‘namespace’ column is introduced, Conductor can allow the duplicate names if their namespaces are different. For example, when users create workflow/task definitions, ‘namespace’ for those definitions can be populated by extracting data from the request contexts. For backward compatibility, if a namespace is not given, a default namespace can be used.

Thank you.

**Comment by venkag** [comment]
`namespace` can act as generic way to isolate the resources. Can be used for multi-tenancy and can be used for long term to solve the noisy-neighbor concerns which are part of the road-map.

https://github.com/Netflix/conductor/wiki/Roadmap (Usability).

@kishorebanala @aravindanr @apanicker-nflx

**Reply by kishorebanala** [reply]
@wonjvmware @venkag Thanks for opening this discussion. We'd like to better understand the goals for `namespace`. 

I.e given the abstract nature of namespace, we'd like to see how you think it'd concretely help your use case. If the namespaces helps to group a set of workflows, have you considered any other options? Vs if the namespaces are aimed at resource isolation, we might need to further hash out the details. Are there any other features you're thinking leveraging namespaces?

**Reply by wonjvmware** [reply]
@kishorebanala Yes. `namespace` are mainly aimed at resource isolation for multi-tenancy. I just realized that multi-tenancy support is in the Conductor road-map. Is there a detailed plan for multi-tenancy support?

**Reply by kishorebanala** [reply]
@wonjvmware Thank you! This is a not a priority at the moment, but please feel free to share your design ideas for multi-tenancy with namespace. If we could agree on a solution that is generic enough to implement within Conductor, we can take it from there.
